### DISCUSSION TITLE
Got weird obeservation data when runing a simulation on intersection env.

### DISCUSSION THREAD

**Discussion by YuqiaoBai** [eleurent/highway-env#265] [Q&A] [answered]
Hi eleurent, thank you for your excellent work! I am working on your project for generating adversarial environment.

I created two IDM vehicles - one as ego vehicle and another one as background vehicle for the adversarial intersection environment, but when I read the obeservation, I found that the position and location data of ego vehicle is unchanged, although the simulation animation worked perfect.

The obeservation looks like beblow:
`[[-2.000000e+00 -3.100000e+01  3.061617e-16  5.000000e+00]
 [-2.000000e+00 -2.600000e+01  3.061617e-16  5.000000e+00]]
[[-2.000000e+00 -3.100000e+01  3.061617e-16  5.000000e+00]
 [-2.000000e+00 -2.100000e+01  3.061617e-16  5.000000e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -1.6000000e+01  1.4163847e-15  5.0000000e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -1.1000000e+01  1.4163847e-15  5.0000000e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -6.0000000e+00  1.4163847e-15  5.0000000e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -3.7682843e+00 -3.1904645e-16 -6.3137311e-01]]
[[-2.000000e+00 -3.100000e+01  3.061617e-16  5.000000e+00]
 [-2.000000e+00 -5.665522e+00  4.066207e-15 -2.099046e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -6.3645701e+00 -1.3996566e-13  9.0085387e-01]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0000000e+00 -4.1061978e+00 -3.0397658e-15  3.6758935e+00]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.0833004e+01 -1.9794830e+00 -4.9995332e+00 -6.8325728e-02]]
[[-2.0000000e+00 -3.1000000e+01  3.0616170e-16  5.0000000e+00]
 [-2.5832979e+01 -1.9944389e+00 -4.9999657e+00 -1.8534983e-02]]`

The simulation animation:

https://user-images.githubusercontent.com/16231913/153379018-a94f5029-7766-451c-ad43-71b5f2ec9a99.mp4

**Comment by eleurent** [comment, marked as answer]
Hi @YuqiaoBai 
I think I would need to look at your code, but if I had to guess, I would say that there is something wrong with how you defined the ego vehicle. It should be created in the environment's` _make_vehicles()` method, set to the `env.vehicle` field, and added to the list of vehicles of the `Road` object.
Here, it looks as though the observer vehicle (1st in the observation, and set to `env.vehicle` by default) was created at some point, but is not part of `road.vehicles`, since it is not being simulated or displayed.

**Comment by YuqiaoBai** [comment]
Thanks for your answer, it helped me a lot. However I faced a new problem which confused me.
Since I need the ids of vehicles, I read the information about vehicles from vehicle.position rather than observation, but the results are diffrent.
below shows the result of a single frame， features of observation are ["x", "y", "vx", "vy"].

```python
original: [ 2. 26.] original: [ -2. -36.] original: [-36. 2.] original: [28. -2.] original: [38.18281435 -2. ] original: [ 2. 37.86826989] 
[[-2.0000000e+00 -4.1000000e+01 3.0616170e-16 5.0000000e+00] 
[-2.0000000e+00 -3.6000000e+01 3.0616170e-16 5.0000000e+00] 
[ 2.8000000e+01 -2.0000000e+00 -3.0000000e+00 3.6739403e-16]
 [ 3.8182816e+01 -2.0000000e+00 -1.4013666e+00 1.7161791e-16]
 [-3.6000000e+01 2.0000000e+00 5.0000000e+00 -2.9605948e-16] 
[ 2.0000000e+00 2.6000000e+01 3.0616170e-16 -5.0000000e+00]]
```

The brief codes of my script:

```python
def main():
env = gym.make('intersection-v1')
env = Monitor(env, directory="videos/intersection/", force=True, video_callable=lambda e: True)
obs = env.reset()
for _ in range(11):
    obs, reward, done, info = env.step(None)
    for vehicle in env.road.vehicles:
        # print("converted:", vehicle.lane.position(longitudinal=vehicle.position[0], lateral=vehicle.position[1]))
        #print("converted:", vehicle.lane.local_coordinates(vehicle.position))
        print("original:", vehicle.position)
    print(obs)
    env.render()
return env
```


Could you help me?

**Comment by eleurent** [comment]
From the observation:
* -2, 41 is not in the original positions
* -2, -36 is original position index 1
* 28, -2 is original position index 3
* 38.18, -2 is original position index 4
* -36, 2 is original position index 2
* 2, 26 is original position index 0

So only the first row of the observation [-2, 41] does not match the remaining original position at index 5: [2, 37.86].
In the kinematics observation, this first row corresponds to the ego vehicle, which again makes me think that there is something wrong in how it is defined / reset?

Would you mind sharing your code, so I can have a look?

**Reply by YuqiaoBai** [reply]
Ok I see the problem and fixed it! thanks for your patience!

Another quick question:
Since all vehicles are controlled by IDM model so It difficult to figure out which one is ego vehicle.
I need to set the ego vehicle to a unique color.
I tried `VehicleGraphics.display(ego_vehicle, surface).color = "PURPLE"` but it didnt work.

**Reply by eleurent** [reply]
You can just do `vehicle.color = ‘PURPLE’` when creating the ego_vehicle.
